lib/src/embedded/compiler/message-transformer.ts

```diff
@@ -0,0 +1,222 @@
+ * at the message level are encapsulated here and reported as ProtocolErrors.
+ */
+export class MessageTransformer {
+  // The decoded messages and protocol errors are written to these Subjects.
+  // They are publicly exposed as readonly Observables to prevent memory leaks.
+  private readonly readInternal$ = new Subject<OutboundMessage>();
+  private readonly protocolErrorInternal$ = new Subject<ProtocolError>();
+
+  /**
+   * The OutboundMessages (decoded from protocol buffers). These are guaranteed
+   * to be valid, since invalid messages are reported as ProtocolErrors.
+   */
+  readonly read$ = from(this.readInternal$);
+
+  /**
+   * Receives InboundMessages and encodes them into protocol buffers. Errors
+   * encountered during encoding are reported as ProtocolErrors.
+   */
+  readonly write$ = new Subject<InboundMessage>();
+
+  /** The ProtocolErrors encountered while encoding/decoding messages. */
+  readonly protocolError$ = from(this.protocolErrorInternal$);
+
+  constructor(
+    private readonly rawRead$: Observable<Buffer>,
+    private readonly rawWrite$: Subject<Buffer>
+  ) {
+    this.rawRead$.subscribe(buffer => this.decode(buffer));
+    this.write$.subscribe(message => this.encode(message));
+  }
+
+  // Decodes a buffer into an OutboundMessage.
+  private decode(buffer: Buffer) {
+    let message;
+    try {
+      message = toMessage(buffer);
+    } catch (error) {
+      this.protocolErrorInternal$.next(error);
+    }
+    this.readInternal$.next(message);
+  }
+
+  // Encodes an InboundMessage into a protocol buffer.
+  private encode(message: InboundMessage) {
+    let buffer;
+    try {
+      buffer = toBuffer(message);
+    } catch (error) {
+      this.protocolErrorInternal$.next(error);
+    }
+    this.rawWrite$.next(buffer);
+  }
+
+  /** Cleans up all Observables. */
+  close() {
+    this.readInternal$.complete();
+    this.write$.complete();
+    this.protocolErrorInternal$.complete();
+  }
+}
+
+// Encodes an InboundMessage into a protocol buffer. Throws a ProtocolError
+// if any errors are encountered.
+function toBuffer(message: InboundMessage): Buffer {
+  try {
+    return Buffer.from(message.serializeBinary());
+  } catch (error) {
+    throw protocolError(
+      ProtocolError.ErrorType.INTERNAL,
+      'Could not serialize buffer'
+    );
+  }
+}
+
+// Decodes an OutboundMessage from a protocol buffer. Throws a ProtocolError
+// if any errors are encountered.
+function toMessage(buffer: Buffer): OutboundMessage {
+  let message;
+
+  try {
+    message = OutboundMessage.deserializeBinary(buffer);
+  } catch (error) {
+    throw protocolError(ProtocolError.ErrorType.PARSE, 'Invalid buffer');
+  }
+
+  // Ensure that the InboundMessage contains a valid message, and that all
+  // mandatory fields are populated.
+  switch (message.getMessageCase()) {
+    case OutboundMessage.MessageCase.COMPILERESPONSE: {
+      const response = message.getCompileresponse()!;
+      if (!response.getId()) {
```

Comment:
There's some tricky stuff going on here. Scalar values in protocol buffers (including ints and even strings) are never null. If the message doesn't include an explicit value, the message will use a [default value](https://developers.google.com/protocol-buffers/docs/proto3#default). This means there's no way to distinguish between a message passing an explicit `0` value for an ID and leaving the ID out entirely. Since the embedded protocol allows 0 as an ID, this effectively means that no message is inherently invalid because it doesn't include an ID. The good news is that this means you have to do a lot less manual checking in this method!

Also, be careful of relying on truthiness in JavaScript—it can behave in unintuitive ways. For example, `0` is falsey, so this will currently throw an error if the ID is set to `0` (explicitly or implicitly). One of those annoying corner cases!